### PR TITLE
8276096: Simplify Unsafe.{load|store}Fence fallbacks by delegating to fullFence

### PR DESCRIPTION
**Pull request by shipilev**
`Unsafe.{load|store}Fence` falls back to `unsafe.cpp` for `OrderAccess::{acquire|release}Fence()`. It seems too heavy-handed (useless?) to call to runtime for a single memory barrier. We can simplify the native `Unsafe` interface by falling back to `fullFence` when `{load|store}Fence` intrinsics are not available. This would be similar to what `Unsafe.{loadLoad|storeStore}Fences` do. 

This is the behavior of these intrinsics now, on x86_64, using benchmarks from JDK-8276054:

```
Benchmark          Mode  Cnt  Score   Error  Units

# Default
Single.acquire     avgt    3   0.407 ± 0.060  ns/op
Single.full        avgt    3   4.693 ± 0.005  ns/op
Single.loadLoad    avgt    3   0.415 ± 0.095  ns/op
Single.plain       avgt    3   0.406 ± 0.002  ns/op
Single.release     avgt    3   0.408 ± 0.047  ns/op
Single.storeStore  avgt    3   0.408 ± 0.043  ns/op

# -XX:DisableIntrinsic=_storeFence
Single.acquire     avgt    3   0.408 ± 0.016  ns/op
Single.full        avgt    3   4.694 ± 0.002  ns/op
Single.loadLoad    avgt    3   0.406 ± 0.002  ns/op
Single.plain       avgt    3   0.406 ± 0.001  ns/op
Single.release     avgt    3   4.694 ± 0.003  ns/op <--- upgraded to full
Single.storeStore  avgt    3   4.690 ± 0.005  ns/op <--- upgraded to full

# -XX:DisableIntrinsic=_loadFence
Single.acquire     avgt    3   4.691 ± 0.001  ns/op <--- upgraded to full
Single.full        avgt    3   4.693 ± 0.009  ns/op
Single.loadLoad    avgt    3   4.693 ± 0.013  ns/op <--- upgraded to full
Single.plain       avgt    3   0.408 ± 0.072  ns/op
Single.release     avgt    3   0.415 ± 0.016  ns/op
Single.storeStore  avgt    3   0.416 ± 0.041  ns/op

# -XX:DisableIntrinsic=_fullFence
Single.acquire     avgt    3   0.406 ± 0.014  ns/op
Single.full        avgt    3  15.836 ± 0.151  ns/op <--- calls runtime
Single.loadLoad    avgt    3   0.406 ± 0.001  ns/op
Single.plain       avgt    3   0.426 ± 0.361  ns/op
Single.release     avgt    3   0.407 ± 0.021  ns/op
Single.storeStore  avgt    3   0.410 ± 0.061  ns/op

# -XX:DisableIntrinsic=_fullFence,_loadFence
Single.acquire     avgt    3  15.822 ± 0.282  ns/op <--- upgraded, calls runtime
Single.full        avgt    3  15.851 ± 0.127  ns/op <--- calls runtime
Single.loadLoad    avgt    3  15.829 ± 0.045  ns/op <--- upgraded, calls runtime
Single.plain       avgt    3   0.406 ± 0.001  ns/op
Single.release     avgt    3   0.414 ± 0.156  ns/op
Single.storeStore  avgt    3   0.422 ± 0.452  ns/op

# -XX:DisableIntrinsic=_fullFence,_storeFence
Single.acquire     avgt    3   0.407 ± 0.016  ns/op
Single.full        avgt    3  15.347 ± 6.783  ns/op <--- calls runtime
Single.loadLoad    avgt    3   0.406 ± 0.001  ns/op
Single.plain       avgt    3   0.406 ± 0.002  ns/op 
Single.release     avgt    3  15.828 ± 0.019  ns/op <--- upgraded, calls runtime
Single.storeStore  avgt    3  15.834 ± 0.045  ns/op <--- upgraded, calls runtime

# -XX:DisableIntrinsic=_fullFence,_loadFence,_storeFence
Single.acquire     avgt    3  15.838 ± 0.030  ns/op <--- upgraded, calls runtime
Single.full        avgt    3  15.854 ± 0.277  ns/op <--- calls runtime
Single.loadLoad    avgt    3  15.826 ± 0.160  ns/op <--- upgraded, calls runtime
Single.plain       avgt    3   0.406 ± 0.003  ns/op
Single.release     avgt    3  15.838 ± 0.019  ns/op <--- upgraded, calls runtime
Single.storeStore  avgt    3  15.844 ± 0.104  ns/op <--- upgraded, calls runtime
```

Additional testing:
 - [x] Linux x86_64 fastdebug `tier1`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276096](https://bugs.openjdk.java.net/browse/JDK-8276096): Simplify Unsafe.{load|store}Fence fallbacks by delegating to fullFence

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to e2c623bef59b40a6115fba1bdae312bc59c2cacd
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to e2c623bef59b40a6115fba1bdae312bc59c2cacd
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6149/head:pull/6149` \
`$ git checkout pull/6149`

Update a local copy of the PR: \
`$ git checkout pull/6149` \
`$ git pull https://git.openjdk.java.net/jdk pull/6149/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6149`

View PR using the GUI difftool: \
`$ git pr show -t 6149`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6149.diff">https://git.openjdk.java.net/jdk/pull/6149.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/vmIntrinsics.hpp [hunk 1]
```diff
@@ -517,13 +517,13 @@ class methodHandle;
...
+  do_intrinsic(_loadFence,                jdk_internal_misc_Unsafe,     loadFence_name, loadFence_signature,           F_R)      \
    do_name(     loadFence_name,                                         "loadFence")                                             \
    do_alias(    loadFence_signature,                                    void_method_signature)                                   \
-  do_intrinsic(_storeFence,               jdk_internal_misc_Unsafe,     storeFence_name, storeFence_signature,         F_RN)     \
+  do_intrinsic(_storeFence,               jdk_internal_misc_Unsafe,     storeFence_name, storeFence_signature,         F_R)      \
    do_name(     storeFence_name,                                        "storeFence")                                            \
    do_alias(    storeFence_signature,                                   void_method_signature)                                   \
-  do_intrinsic(_fullFence,                jdk_internal_misc_Unsafe,     fullFence_name, fullFence_signature,           F_RN)     \
+  do_intrinsic(_fullFence,                jdk_internal_misc_Unsafe,     fullFence_name, fullFence_signature,           F_R)      \
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/vmIntrinsics.hpp:526]: Why did you drop the N from F_RN? AFAICS the fullFence method is still native.

**Comment by dholmes-ora** [comment]
I'm not quite seeing the motivation here. Your claim is that the non-intrinsic implementations involve a native call and so that is too expensive; yet the new code still relies on the fullFence being intrinsified else it is still a native call and a heavier barrier. If these fences were intrinisified piecemeal then perhaps this is an issue on some platform, but is that really the case? If you intrinsified one wouldn't you intrinsify all?

File: src/hotspot/share/classfile/vmIntrinsics.hpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/classfile/vmIntrinsics.hpp]: Good spot! That's indeed incorrect, fixed in new commit. I am surprised `CheckIntrinsics` did not found this discrepancy.  I believe "native" flags are not checked at all? For example, existing `_hashCode` intrinsic is also `F_R`, while it covers the native `java.lang.Object::hashCode`. I try to beef up those checks separately.

**Comment by shipilev** [comment]
> I'm not quite seeing the motivation here. Your claim is that the non-intrinsic implementations involve a native call and so that is too expensive; yet the new code still relies on the fullFence being intrinsified else it is still a native call and a heavier barrier. If these fences were intrinisified piecemeal then perhaps this is an issue on some platform, but is that really the case? If you intrinsified one wouldn't you intrinsify all?

Yes, that was not clear, sorry. For current platforms, it is mostly a maintenance cleanup to shrink the unnecessary Unsafe interfaces: if we disable the `acquireFence` intrinsic, we don't need to call into native fallback (which would be excessive), instead we can just go to Java-level fallback (which would also be faster). 

I am looking at the cases where we would like to only intrinsify `fullFence`, for example for Zero interpreter. Instead of handling all three flavors of fences, we can get the majority of performance win by only drilling the interpreter-entry-intrinsic hole for `fullFence`, and let everything else handled at Java level.

File: src/hotspot/share/classfile/vmIntrinsics.hpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/classfile/vmIntrinsics.hpp]: This `CheckIntrinsics` oddity is handled by #6187.

**Review by dholmes-ora — APPROVED** [review]

**Comment by shipilev** [comment]
Thanks!

/integrate

**Pull request closed without merging**